kt/godot-runtime/src/main/kotlin/godot/core/Variants.kt

```diff
@@ -1,337 +1,285 @@
+fun getVariantType(transform: Transform) = VariantType.TRANSFORM to transform
+fun getVariantType(variantArray: VariantArray) = VariantType.ARRAY to variantArray
+fun getVariantType(ktObject: KtObject) = VariantType.OBJECT to ktObject
+
+var ByteBuffer.bool: Boolean
+    get() = int == 1
+    set(value) {
+        putInt(if (value) 1 else 0)
+    }
+
+private var ByteBuffer.vector2: Vector2
+    get() = Vector2(float.toRealT(), float.toRealT())
+    set(value) {
+        putFloat(value.x.toFloat())
+        putFloat(value.y.toFloat())
+    }
+
+private var ByteBuffer.vector3: Vector3
+    get() = Vector3(float.toRealT(), float.toRealT(), float.toRealT())
+    set(value) {
+        putFloat(value.x.toFloat())
+        putFloat(value.y.toFloat())
+        putFloat(value.z.toFloat())
+    }
+
+private var ByteBuffer.basis: Basis
+    get() = Basis().also {
+        it._x = vector3
+        it._y = vector3
+        it._z = vector3
+    }
+    set(value) {
+        vector3 = value._x
+        vector3 = value._y
+        vector3 = value._z
+    }
+
+
+internal fun Any.encode(type: VariantType, buffer: ByteBuffer) {
+    buffer.putInt(type.ordinal)
+    type.toGodot(buffer, this)
+}
 
-        private fun Wire.Vector3.toKVector3(): Vector3 {
-            return Vector3(x, y, z)
-        }
+internal fun parse(buffer: ByteBuffer): Any {
+    val int = buffer.int
+    return VariantType.values()[int].toKotlin(buffer)
+}
 
-        private fun Wire.Basis.toKBasis(): Basis {
-            // write to the internal values directly
-            return Basis().also {
-                it._x = Vector3(this.x.x, this.x.y, this.x.z)
-                it._y = Vector3(this.y.x, this.y.y, this.y.z)
-                it._z = Vector3(this.z.x, this.z.y, this.z.z)
+inline fun <reified T> Any.asObject(): T = this as T
+
+enum class VariantType (
+        internal val toKotlin: (ByteBuffer) -> Any,
+        internal val toGodot: (ByteBuffer, any: Any) -> Unit
+) {
+    NIL(
+            { buffer: ByteBuffer -> Unit },
+            { buffer: ByteBuffer, any: Any -> Unit }
+    ),
+
+    // atomic types
+    BOOL(
+            { buffer: ByteBuffer -> buffer.bool },
+            { buffer: ByteBuffer, any: Any -> buffer.bool = any as Boolean }
+    ),
+    LONG(
+            { buffer: ByteBuffer -> buffer.long },
+            { buffer: ByteBuffer, any: Any -> buffer.putLong(any as Long) }
+    ),
+    DOUBLE(
+            { buffer: ByteBuffer -> buffer.double},
+            { buffer: ByteBuffer, any: Any -> buffer.putDouble(any as Double) }
+    ),
+    STRING(
+            { buffer: ByteBuffer ->
+                val stringSize = buffer.int
+                val charArray = kotlin.ByteArray(stringSize)
+                buffer.get(charArray, 0, stringSize)
+                kotlin.text.String(charArray, kotlin.text.Charsets.UTF_8)
+            },
+            { buffer: ByteBuffer, any: Any ->
+                any as String
```

Comment:
this is a bit weird, maybe use `require`?

Reply:
Here, the check is not necessary, by design, we know the type we have as it is linked to the enum.

Reply:
Yep, it is just an unusual way to do a smart cast, may be do 

```kotlin
val stringBytes = (any as String).encodeToByteArray()
```

Reply:
or if you must really do that - then I suggest doing a `require(any as String)` - it's not really about the check but more of the readability.

Reply:
If it is only readability purpose, why not only renaming any parameter ? I mean, yes it is unusual, but it is the most optimized way, and what we do here is somewhat unusual I think ^^